File: pandoc-runner/src/UnixSocketServer.ts
import * as net from 'net';
import * as fs from 'fs';
import { Logger } from './types';
import { DefaultLogger } from './Logger';

/**
 * Unixソケット経由での通信を管理するクラス
 */
export class UnixSocketServer {
  private static readonly DEFAULT_SOCKET_PATH = '/tmp/pandoc-runner.sock';
  
  private socketPath: string;
  private logger: Logger;
  private server: net.Server | null = null;
  private running = false;

  constructor(socketPath?: string, logger?: Logger) {
    this.socketPath = socketPath || UnixSocketServer.DEFAULT_SOCKET_PATH;
    this.logger = logger || new DefaultLogger();
  }

  /**
   * サーバーを開始する
   * @param requestHandler リクエストを処理するコールバック関数
   */
  async start(requestHandler: (requestLine: string) => Promise<string>): Promise<void> {
    return new Promise((resolve, reject) => {
      this.cleanupSocket();
      
      this.server = net.createServer();
      this.running = true;

      this.server.on('connection', (socket: net.Socket) => {
        this.handleClient(socket, requestHandler);
      });

      this.server.on('error', (error: Error) => {
        this.logger.error(`Server error: ${error.message}`);
        reject(error);
      });

      this.server.listen(this.socketPath, () => {
        this.logger.info(`UnixSocket server started at ${this.socketPath}`);
        resolve();
      });
    });
  }

  /**
   * サーバーを停止する
   */
  async stop(): Promise<void> {
    if (!this.running || !this.server) {
      return;
    }

    return new Promise((resolve) => {
      this.running = false;
      
      this.server!.close(() => {
        this.cleanupSocket();
        this.logger.info('UnixSocket server stopped');
        resolve();
      });
    });
  }

  /**
   * サーバーが実行中かどうか
   */
  isRunning(): boolean {
    return this.running;
  }

  /**
   * ソケットパスを取得
   */
  getSocketPath(): string {
    return this.socketPath;
  }

  /**
   * クライアントからの接続を処理する
   * @param socket クライアントソケット
   * @param requestHandler リクエストハンドラー
   */
  private handleClient(
    socket: net.Socket, 
    requestHandler: (requestLine: string) => Promise<string>
  ): void {
    let buffer = '';

    socket.on('data', async (data: any) => {
      buffer += data.toString();
      
      // 改行で区切られたリクエストを処理
      const lines = buffer.split('\n');
      buffer = lines.pop() || ''; // 最後の不完全な行をバッファに保持

      for (const line of lines) {
        if (line.trim()) {
          try {
            this.logger.debug(`Received request: ${line.trim()}`);
            const response = await requestHandler(line.trim());
            
            socket.write(response + '\n');
            this.logger.debug(`Sent response: ${response}`);
          } catch (error) {
            this.logger.error(`Error handling request: ${error}`);
            const errorResponse = JSON.stringify({
              success: false,
              error: { message: `${error}`, code: 'INTERNAL_ERROR' },
              timestamp: new Date().toISOString()
            });
            socket.write(errorResponse + '\n');
          }
        }
      }
    });

    socket.on('error', (error: Error) => {
      this.logger.error(`Client socket error: ${error.message}`);
    });

    socket.on('close', () => {
      this.logger.debug('Client disconnected');
    });
  }

  /**
   * ソケットファイルをクリーンアップする
   */
  private cleanupSocket(): void {
    try {
      if (fs.existsSync(this.socketPath)) {
        fs.unlinkSync(this.socketPath);
      }
    } catch (error) {
      this.logger.warn(`Failed to cleanup socket: ${error}`);
    }
  }
}

File: pandoc-runner-rb/lib/unix_socket_server.rb
# frozen_string_literal: true

require 'socket'
require 'logger'

# UnixSocketServerクラス
# Unixソケット経由での通信を管理する
class UnixSocketServer
  DEFAULT_SOCKET_PATH = '/tmp/pandoc-runner.sock'

  attr_reader :socket_path, :logger

  def initialize(socket_path: DEFAULT_SOCKET_PATH, logger: nil)
    @socket_path = socket_path
    @logger = logger || Logger.new($stdout)
    @server = nil
    @running = false
  end

  # サーバーを開始する
  def start(&block)
    cleanup_socket
    @server = UNIXServer.new(@socket_path)
    @running = true

    @logger.info("UnixSocket server started at #{@socket_path}")

    accept_connections(&block)
  ensure
    stop
  end

  # サーバーを停止する
  def stop
    return unless @running

    @running = false
    @server&.close
    cleanup_socket

    @logger.info('UnixSocket server stopped')
  end

  # サーバーが実行中かどうか
  def running?
    @running
  end

  private

  # 接続を受け入れる
  def accept_connections(&block)
    while @running
      begin
        ready = IO.select([@server], nil, nil, 1)
        next unless ready

        client = @server.accept
        handle_client(client, &block)
      rescue StandardError => e
        @logger.error("Error in accept_connections: #{e.message}")
        @logger.error(e.backtrace)
      end
    end
  end

  # クライアントからの接続を処理する
  # @param client [UNIXSocket] クライアントソケット
  def handle_client(client, &block)
    Thread.new do
      request_line = client.gets
      return unless request_line

      request_line = request_line.strip
      @logger.debug("Received request: #{request_line}")

      response = block.call(request_line) if block

      if response
        client.puts(response)
        @logger.debug("Sent response: #{response}")
      end
    rescue StandardError => e
      @logger.error("Error handling client: #{e.message}")
      error_response = { error: e.message }.to_json
      client.puts(error_response)
    ensure
      client.close
    end
  end

  # ソケットファイルをクリーンアップする
  def cleanup_socket
    File.unlink(@socket_path) if File.exist?(@socket_path)
  rescue StandardError => e
    @logger.warn("Failed to cleanup socket: #{e.message}")
  end
end
